video/out/drm_common.c

```diff
@@ -497,19 +499,79 @@ static bool setup_mode(struct kms *kms, const char *mode_spec)
     return false;
 }
 
-static int open_card(int card_no)
+static int open_card_path(const char *path)
 {
-    char card_path[128];
-    snprintf(card_path, sizeof(card_path), DRM_DEV_NAME, DRM_DIR_NAME, card_no);
-    return open(card_path, O_RDWR | O_CLOEXEC);
+    return open(path, O_RDWR | O_CLOEXEC);
+}
+
+static char *get_primary_device_path(struct mp_log *log, int *card_no)
+{
+    drmDevice *devices[DRM_MAX_MINOR] = { 0 };
+    int card_count = drmGetDevices(devices,
+                                   MP_ARRAY_SIZE(devices));
+    char *device_path = NULL;
+    bool card_no_given = (*card_no >= 0);
+
+    if (card_count < 0) {
+        mp_err(log, "Listing DRM devices with drmGetDevices failed! (%s)\n",
+               mp_strerror(errno));
+    }
+
+    if (card_no_given && *card_no > (card_count - 1)) {
+        mp_err(log, "Card number %d given too high! %d devices located.\n",
+               *card_no, card_count);
+        goto err;
+    }
+
+    for (int i = card_no_given ? *card_no : 0; i < card_count; i++) {
+        drmDevice *dev = devices[i];
+
+        if (!dev || !(dev->available_nodes & (1 << DRM_NODE_PRIMARY))) {
+            if (card_no_given) {
+                mp_err(log,
+                       "DRM card number %d given, yet it does not have "
+                       "a primary node!\n", i);
+                break;
+            }
+
+            continue;
+        }
+
+        char *primary_node_path = dev->nodes[DRM_NODE_PRIMARY];
+        if (!primary_node_path || !primary_node_path[0]) {
```

Comment:
Similarly - if `DRM_NODE_PRIMARY` is set, then `primary_node_path` is guaranteed to be a sane string. Please drop the NULL and empty string checks.